test_conformance/relationals/test_comparisons_fp.cpp

```diff
@@ -0,0 +1,662 @@
+    : context(context), device(device), queue(queue), fnName(fn), opName(op)
+{
+    // hardcoded for now, to be changed into typeid().name solution in future
+    // for now C++ spec doesn't guarantee human readable type name
+
+    eqTypeNames = { { kHalf, "short" },
+                    { kFloat, "int" },
+                    { kDouble, "long" } };
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T>
+void RelationalsFPTest::generate_equiv_test_data(T* outData,
+                                                 const unsigned int& vecSize,
+                                                 const bool& alpha,
+                                                 const RelTestParams<T>& param,
+                                                 const MTdata& d)
+{
+    unsigned int i;
+
+    generate_random_data(param.dataType, vecSize * TEST_SIZE, d, outData);
+
+    // Fill the first few vectors with NAN in each vector element (or the second
+    // set if we're alpha, so we can test either case)
+    if (alpha) outData += vecSize * vecSize;
+    for (i = 0; i < vecSize; i++)
+    {
+        outData[0] = param.nan;
+        outData += vecSize + 1;
+    }
+    // Make sure the third set is filled regardless, to test the case where both
+    // have NANs
+    if (!alpha) outData += vecSize * vecSize;
+    for (i = 0; i < vecSize; i++)
+    {
+        outData[0] = param.nan;
+        outData += vecSize + 1;
+    }
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T, typename U>
+void RelationalsFPTest::verify_equiv_values(const unsigned int& vecSize,
+                                            const T* const inDataA,
+                                            const T* const inDataB,
+                                            U* const outData,
+                                            const VerifyFunc<T>& verifyFn)
+{
+    unsigned int i;
+    int trueResult;
+    bool result;
+
+    trueResult = (vecSize == 1) ? 1 : -1;
+    for (i = 0; i < vecSize; i++)
+    {
+        result = verifyFn(inDataA[i], inDataB[i]);
+        outData[i] = result ? trueResult : 0;
+    }
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T>
+int RelationalsFPTest::test_equiv_kernel(const unsigned int& vecSize,
+                                         const RelTestParams<T>& param,
+                                         const MTdata& d)
+{
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+    clMemWrapper streams[4];
+    T inDataA[TEST_SIZE * 16], inDataB[TEST_SIZE * 16];
+
+    // support half, float, double equivalents - otherwise assert
+    typedef typename std::conditional<
+        (sizeof(T) == sizeof(short)), short,
+        typename std::conditional<(sizeof(T) == sizeof(int)), int,
+                                  long>::type>::type U;
```

Comment:
It might be more portable to test for the fixed width integer types (`int16_t` etc.)

Reply:
Done.